util/linuxfw/linuxfw.go

```diff
@@ -2,10 +2,250 @@
+	Accept MatchDecision = iota
+	Masq
+)
+
+// The following bits are added to packet marks for Tailscale use.
+//
+// We tried to pick bits sufficiently out of the way that it's
+// unlikely to collide with existing uses. We have 4 bytes of mark
+// bits to play with. We leave the lower byte alone on the assumption
+// that sysadmins would use those. Kubernetes uses a few bits in the
+// second byte, so we steer clear of that too.
+//
+// Empirically, most of the documentation on packet marks on the
+// internet gives the impression that the marks are 16 bits
+// wide. Based on this, we theorize that the upper two bytes are
+// relatively unused in the wild, and so we consume bits 16:23 (the
+// third byte).
+//
+// The constants are in the iptables/iproute2 string format for
+// matching and setting the bits, so they can be directly embedded in
+// commands.
+const (
+	// The mask for reading/writing the 'firewall mask' bits on a packet.
+	// See the comment on the const block on why we only use the third byte.
+	//
+	// We claim bits 16:23 entirely. For now we only use the lower four
+	// bits, leaving the higher 4 bits for future use.
+	TailscaleFwmarkMask    = "0xff0000"
+	TailscaleFwmarkMaskNeg = "0xff00ffff"
+	TailscaleFwmarkMaskNum = 0xff0000
+
+	// Packet is from Tailscale and to a subnet route destination, so
+	// is allowed to be routed through this machine.
+	TailscaleSubnetRouteMark    = "0x40000"
+	TailscaleSubnetRouteMarkNum = 0x40000
+	// This one is same value but padded to even number of digit, so
+	// hex decoding can work correctly.
+	TailscaleSubnetRouteMarkHexStr = "0x040000"
+
+	// Packet was originated by tailscaled itself, and must not be
+	// routed over the Tailscale network.
+	//
+	// Keep this in sync with tailscaleBypassMark in
+	// net/netns/netns_linux.go.
+	TailscaleBypassMark    = "0x80000"
+	TailscaleBypassMarkNum = 0x80000
+)
+
+// routeTable is a Linux routing table: both its name and number.
+// See /etc/iproute2/rt_tables.
+type RouteTable struct {
+	Name string
+	Num  int
+}
+
+// ipCmdArg returns the string form of the table to pass to the "ip" command.
+func (rt RouteTable) IpCmdArg() string {
+	if rt.Num >= 253 {
+		return rt.Name
+	}
+	return strconv.Itoa(rt.Num)
+}
+
+var routeTableByNumber = map[int]RouteTable{}
+
+func newRouteTable(name string, num int) RouteTable {
+	rt := RouteTable{name, num}
+	routeTableByNumber[num] = rt
+	return rt
+}
+
+func MustRouteTable(num int) RouteTable {
+	rt, ok := routeTableByNumber[num]
+	if !ok {
+		panic(fmt.Sprintf("unknown route table %v", num))
+	}
+	return rt
+}
+
+// errCode extracts and returns the process exit code from err, or
+// zero if err is nil.
+func errCode(err error) int {
+	if err == nil {
+		return 0
+	}
+	var e *exec.ExitError
+	if ok := errors.As(err, &e); ok {
+		return e.ExitCode()
+	}
+	s := err.Error()
+	if strings.HasPrefix(s, "exitcode:") {
+		code, err := strconv.Atoi(s[9:])
+		if err == nil {
+			return code
+		}
+	}
+	return -42
+}
```

Comment:
1. I think in linuxfw package we only care about if it's 1 (which maps to chain doesn't exist in our use cases)
2. This I believe is a tool that's used in router, to get errorCode from different types of errors . I doubt it would mean a lot for us to doc these package dependent error codes. We look for a errorCode == 2 from `osCommandRunner` in `router_linux.go`, errorCode == 1 from `nftables.ClearChain` and some other code in runGroup. These values doesn't seem necessary for us to record, and they aren't really related either. We might have another 2 from another package one day.

Reply:
Also I copied rather than moved the errorCode implementation from `wgengine/router/runner.go` to linuxfw package and use it from import. It doesn't make sense to import this util function from another package. It's not a linuxfw specific thing. Adding this to a third util package would make some sense to me if you have a good place in mind,